controllers/infra/capability/infra_capability_controller.go

```diff
@@ -0,0 +1,163 @@
+	)
+
+	r := NewReconciler(
+		ctx,
+		mgr.GetClient(),
+		ctrl.Log.WithName("controllers").WithName(controllerName),
+		record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+		ctx.VMProvider,
+	)
+
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	cache, err := pkgmgr.NewNamespacedCacheForObject(
+		mgr,
+		&ctx.SyncPeriod,
+		controlledType,
+		WCPClusterCapabilitiesNamespace)
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(source.Kind(
+		cache,
+		controlledType,
+		&handler.TypedEnqueueRequestForObject[*corev1.ConfigMap]{},
+		predicate.TypedFuncs[*corev1.ConfigMap]{
+			CreateFunc: func(e event.TypedCreateEvent[*corev1.ConfigMap]) bool {
+				return e.Object.GetName() == WCPClusterCapabilitiesConfigMapName
+			},
+			UpdateFunc: func(e event.TypedUpdateEvent[*corev1.ConfigMap]) bool {
+				return e.ObjectOld.GetName() == WCPClusterCapabilitiesConfigMapName
+			},
+			DeleteFunc: func(e event.TypedDeleteEvent[*corev1.ConfigMap]) bool {
+				return false
+			},
+			GenericFunc: func(e event.TypedGenericEvent[*corev1.ConfigMap]) bool {
+				return false
+			},
+		},
+		kubeutil.TypedResourceVersionChangedPredicate[*corev1.ConfigMap]{},
+	))
+}
+
+func NewReconciler(
+	ctx context.Context,
+	client client.Client,
+	logger logr.Logger,
+	recorder record.Recorder,
+	vmProvider providers.VirtualMachineProviderInterface) *Reconciler {
+	return &Reconciler{
+		Context:    ctx,
+		Client:     client,
+		Logger:     logger,
+		Recorder:   recorder,
+		VMProvider: vmProvider,
+	}
+}
+
+type Reconciler struct {
+	client.Client
+	Context    context.Context
+	Logger     logr.Logger
+	Recorder   record.Recorder
+	VMProvider providers.VirtualMachineProviderInterface
+}
+
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx = pkgcfg.JoinContext(ctx, r.Context)
+
+	if req.Name == WCPClusterCapabilitiesConfigMapName && req.Namespace == WCPClusterCapabilitiesNamespace {
+		return ctrl.Result{}, r.reconcileWcpClusterCapabilitiesConfig(ctx, req)
+	}
+
+	r.Logger.Error(nil, "Reconciling unexpected object", "req", req.NamespacedName)
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) reconcileWcpClusterCapabilitiesConfig(ctx context.Context, req ctrl.Request) error {
+	r.Logger.Info("Reconciling WCP Cluster Capabilities Config", "configMap", req.NamespacedName)
+
+	cm := &corev1.ConfigMap{}
+	if err := r.Get(ctx, req.NamespacedName, cm); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	isTKGMultipleCL := isTKGMultipleCLSupported(cm.Data)
+
+	// The SetContext call impacts the configuration available to contexts throughout the process,
+	// not just *this* context or its children. Please refer to the pkg/config package for more information
+	// on SetContext and its behavior.
+	if pkgcfg.FromContext(ctx).Features.TKGMultipleCL != isTKGMultipleCL {
```

Comment:
Rewrite as:

```go
pkgcfg.SetContext(ctx, func(config *pkgcfg.Config) {
	config.Features.TKGMultipleCL = isTKGMultipleCLSupported(cm.Data)
})
```

There is no reason to check the value before setting it, it just involves getting two locks instead of one.